File: 0266. Palindrome Permutation.java
// Easy
// Hash Table
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/palindrome-permutation/

import java.util.*;

class Solution {
    // One char can be odd, others should be even
    public boolean canPermutePalindrome(String s) {
        Map<Character, Integer> map = new HashMap<>();
        for (char c : s.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int odd = 0;
        for (int x : map.values()) {
            if (x % 2 != 0)
                odd++;
            if (odd > 1)
                return false;
        }
        return true;
    }
}

File: 0001-0500/0266. Palindrome Permutation.py
# Time:O(n), Space:O(1)

from typing import Counter

class Solution:
    def canPermutePalindrome(self, s: str) -> bool:
        count = Counter(s)
        odd_count = 0
        for freq in count.values():
            if freq % 2 == 1:
                odd_count += 1

        return odd_count <= 1
    